### DISCUSSION TITLE
Log stats per-repo per-run

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21667] [Suggest an Idea]
It would be useful to summarise how many branches created/rebased and PRs created it updated for each run, as well as a count of branches and PRs open. 

Edge cases:

- we exit after automerging so that run would not be an accurate measure of branches/PRs open. Meanwhile the summary from the previous run would be out of date

**Comment by Lucas-C** [comment]
I'd like to start working on such feature.

This post aims at drafting a design / spec around this feature, and is a [request for comments](https://fr.wikipedia.org/wiki/Request_for_comments).

## Goal : exposing the result of a `renovate` run as a parseble output

The idea is to generate a JSON report file at the end of `renovate` execution, so that another tool could post-process this report and send it to a monitoring tool [as metrics](https://openmicroservices.org/schema/metrics/).

**Notes:**
- alternatively, a metrics-publishing lib could be integrated to `renovate`, to allow for continuous metrics publishing as the process runs. While potentially more useful to closely follow `renovate` execution, this would means more coupling with specific metrics standard / format, and a more heavy integration in the code base (_e.g._ dedicated metrics worker ?)
- if the string `-` is provided as the filename, the output will be sent to `renovate` stdout, allowing for better integration with logs aggregation tools

### Configuration
This report-publishing behaviour would be entirely optional. Like all other [configuration options](https://github.com/renovatebot/renovate/blob/master/docs/usage/configuration-options.md), it could be specified as an environment variable, a CLI flag or in the dedicated configuration file.

At the code level, a new "admin" "global" `RenovateBooleanOption` named `--json-report-filepath` will be added in `config/definitions.ts`.

### Metrics collected
- the [existing host statistics](https://github.com/renovatebot/renovate/blob/19.114.0/lib/util/got/stats.ts)
- the number of PR created during the run
- the number of branches created during the run
- the number of branches merged during the run
- the number of repositories analyzed
- the number of dependency descriptor files parsed
- statistics on error cases:
  * number of invalid `renovate.json` files parsed
  * number of invalid dependency descriptor files parsed
  * number of dependencies not found while resolving them
- a count of branches and PRs created by `renovate` and still open / closed

Depending on the complity of exposing those metrics, they may not be included in the 1st version of this feature.

### Code reuse & modification
This feature would take inspiration and extend the existing `util/got/stats` module.
A "state" object containing all the stats should be accessible throughout the app. It is currently defined as module-private "global" variable in `got/stats.ts`, that only tracks HTTP requests performed. The suggested change will be to store a new "StatsRecorder" object along with the logging logic, in the `logger/` package, with new methods added to the `Logger` interface, like `incrMetric(name)`.

**Comment by rarkins** [comment]
Awesome! Thanks for taking the initiative.

> The idea is to generate a JSON report file at the end of `renovate` execution

I would like first to consider/rule out simply incorporating it into the existing logging. Logging in Renovate is fundamentally object/json-based, although the CLI defaults to pretty printing for human consumption. If you intend to consume the logs as a machine, would the existing LOG_FORMAT=json approach suffice?

Although as you already seem aware of, we lean towards making everything configurable anyway, so having both approaches (embedded within logs, or output to a file) is also fine.

> * alternatively, a metrics-publishing lib could be integrated to `renovate`

I'd lean away from this approach unless outputting via CLI or file simply doesn't cut it.

> This report-publishing behaviour would be entirely optional.

Alternatively, we could start by logging the stats we're interested in at debug by default, and then make a form file-based output a second phase. 

> * the number of PR created during the run
> * the number of branches created during the run
> * the number of branches merged during the run

I would prefer to see these per-repo rather than per-run. For the app's case, this could then be used to display stats next to each log file for quicker locating of "interesting" runs. Is there much value for you in seeing them per-run? (i.e. the combined values for potentially hundreds of repos)? It shouldn't be too hard to dump certain (appropriate) stats per-repo and then aggregate them for eventual per-run output though.

**Comment by HenrikPoulsen** [comment]
I would be interested in taking a stab at this, since I have a need for getting some analytics out of renovate. So we can see what is going on in our various repos when it comes to renovate on the macro level.
Though I would likely want more granular information than what is outlined in this issue.
Things such as create/close time of branches and PRs, which prs/branches are open/closed/deleted/updated/unchanged and include which dependency and version the given PR/branch is for etc.
Since we have had repos internally that haven't had any notifications setup on PRs so renovate PRs have been sitting there for months untouched for example.
So I guess the question is if it makes sense to you guys to have this additional information reported as well, or if you feel like that is outside the scope of this issue and/or what renovate should be doing.

**Comment by rarkins** [comment]
I think best if you can list out your requirements within this issue first and then we can confirm if they seem appropriate for logging as stats or not. I think I would need to "see" the idea to be sure.

**Comment by HenrikPoulsen** [comment]
Alright sure.
The things I would like to be able to see is how successful renovate is in our github enterprise organization.
So seeing how many PRs renovate are opening per repo (and for which dependencies and versions the PRs are for), as well as for how long the PRs have been open (creation time).

So that we with that information we could internally make our own renovate global dashboard that shows us which dependencies are outdated in our repos currently and see which repos are stagnating, with PRs that are unattended etc.

**Comment by rarkins** [comment]
Appreciate the high level requirements, but I was thinking more low level in this case, e.g. to determine if what you need can be done elegantly and in a format that's beneficial to most

**Comment by HenrikPoulsen** [comment]
That's a fair point.
So just a quick draft of what it might look like. I'm sure the structure and naming can be improved, but should give some idea of what I was thinking anyway I hope.
```
{
  "startTime": "2022-01-18T13:44:15.993Z",
  "endTime": "2022-01-18T13:48:06.404Z",
  "renovateVersion": "0.0.0-semantic-release",
  "repositories": {
    "henrikp/renovate-stats-tests": {
      "dependencyUpdates": [
        {
          "branch": "renovate/asyncio-nats-client-0.x",
          "createdAt": "2022-01-14T13:32:25Z",
          "datasource": "pypi",
          "depName": "asyncio-nats-client",
          "depCurrentVersion": "0.11.4",
          "depNewVersion": "0.11.5",
          "prNumber": 7,
          "state": "created"
        },
        {
          "branch": "renovate/flask-1.x",
          "createdAt": "2022-01-14T13:32:33Z",
          "datasource": "pypi",
          "depName": "flask",
          "depCurrentVersion": "1.1.2",
          "depNewVersion": "1.1.4",
          "prNumber": 8,
          "state": "automerged"
        },
        {
          "branch": "renovate/lz4-3.x",
          "createdAt": "2022-01-14T13:32:41Z",
          "datasource": "pypi",
          "depName": "lz4",
          "depCurrentVersion": "3.1.3",
          "depNewVersion": "3.1.10",
          "prNumber": 9,
          "state": "unchanged"
        }
      ],
      "startedAt": "2022-01-18T13:46:35.600Z",
      "endedAt": "2022-01-18T13:47:52.163Z"
    }
  },
  "loggerErrors": []
}
```

**Comment by rarkins** [comment]
Seems reasonable, but I think I'd prefer to log repo stats/metadata per-repo than all combined at the end of a run.

**Comment by HenrikPoulsen** [comment]
How would that look in the end? Do you envision it just printing the stats to stdout, along with the existing logging, rather than to a dedicated file?

**Comment by rarkins** [comment]
I think it could be a regular debug message.

**Comment by HenrikPoulsen** [comment]
Ah I see. In my case I would like to extract this data though for all runs. And debug logging is not something we enable by default because it is verbose.
This is why I went with dumping it to a file at the end of the run so that it is easily processed at the end, so we can send the data where we need it to for further processing

**Comment by rarkins** [comment]
In this case it may be better suited to adding to Renovate's repository cache file

**Comment by HenrikPoulsen** [comment]
Nothing would prevent us from doing both of course, since I can understand why people would want that information visible as a debug message as well. Same data for 2 different use cases

**Comment by viceice** [comment]
You can enable debug logging to file, then you get a nice json logfile, which you can easily parse with `jq` or something else.

**Comment by HenrikPoulsen** [comment]
Made a draft PR for what I have so far. So I don't spend too much time on something that might go completely against the renovate grain #13923
Any feedback appreciated
